### ISSUE TITLE
PublicKeyJWK and PrivateKeyJWK are il-defined

### ISSUE THREAD

**Issue by decentralgabe** [w3c-ccg/lds-jws2020#94]
The test vectors call out keys like this:

```
{
  "keypair_0": {
    "id": "#ovsDKYBjFemIy8DVhc-w2LSi8CvXMw2AYDzHj04yxkc",
    "type": "JsonWebKey2020",
    "controller": "did:key:z6Mkf5rGMoatrSj1f4CyvuHBeXJELe9RPdzo2PKGNCKVtZxP",
    "publicKeyJwk": {
      "kty": "OKP",
      "crv": "Ed25519",
      "x": "CV-aGlld3nVdgnhoZK0D36Wk-9aIMlZjZOK2XhPMnkQ"
    },
    "privateKeyJwk": {
      "kty": "OKP",
      "crv": "Ed25519",
      "x": "CV-aGlld3nVdgnhoZK0D36Wk-9aIMlZjZOK2XhPMnkQ",
      "d": "m5N7gTItgWz6udWjuqzJsqX-vksUnxJrNjD5OilScBc"
    }
  }
```

but I cannot find _any_ documentation on how to generate these objects and what fields they may possibly contain. I spent hours banging my head against the wall trying to base64URL decode `d` - trying various libraries - and reading varioius RFCs such as [RFC8037](https://datatracker.ietf.org/doc/html/rfc8037#section-2) which defines the OKP key type. The RFC mentions, in section 2

> The parameter "d" MUST be present for private keys and contain the
      private key encoded using the base64url encoding.  This parameter
      MUST NOT be present for public keys.

Clearly one would assume `d` is a properly base64URL encoded property. but....that's not the case. the standard golang library fails with

```
d := "m5N7gTItgWz6udWjuqzJsqX-vksUnxJrNjD5OilScBc"
decodedD, err := base64.URLEncoding.DecodeString(d)
```

`illegal base64 data at input byte 40`! This is go standard library code. Nothing fancy.

I still cannot find any documentation. I consider this suite to be _functionally broken_ until there is clear authoritative documentation on how to construct and deconstruct these key types.

**Comment by OR13** [comment]
hmm possible the test vectors were generated from a faulty implementation (probably my fault)

**Comment by OR13** [comment]
Can you confirm the same issue occcurs with: 

https://did.key.transmute.industries/generate/ed25519?seed=7d61904c541a11cd8a87e2f22615e5510b5c6a4b45ca63ed55f702e5269c45ec

```
[
  {
    "id": "did:key:z6MkvVSZzJN1UjhY6YmWArSixWTVT2wV8kk51K5HtfDBbYij#z6MkvVSZzJN1UjhY6YmWArSixWTVT2wV8kk51K5HtfDBbYij",
    "type": "JsonWebKey2020",
    "controller": "did:key:z6MkvVSZzJN1UjhY6YmWArSixWTVT2wV8kk51K5HtfDBbYij",
    "publicKeyJwk": {
      "kty": "OKP",
      "crv": "Ed25519",
      "x": "7koNVIt6U4DEJ2R8RzVW2OkOPfVpmZPXB5iS6NJ1pUA"
    },
    "privateKeyJwk": {
      "kty": "OKP",
      "crv": "Ed25519",
      "x": "7koNVIt6U4DEJ2R8RzVW2OkOPfVpmZPXB5iS6NJ1pUA",
      "d": "fWGQTFQaEc2Kh-LyJhXlUQtcaktFymPtVfcC5SacRew"
    }
  },
  {
    "id": "did:key:z6MkvVSZzJN1UjhY6YmWArSixWTVT2wV8kk51K5HtfDBbYij#z6LSombNLmhgBwAW4gb9ffAcCCuHw5sH6AUW9QJSoyrah6th",
    "type": "JsonWebKey2020",
    "controller": "did:key:z6MkvVSZzJN1UjhY6YmWArSixWTVT2wV8kk51K5HtfDBbYij",
    "publicKeyJwk": {
      "kty": "OKP",
      "crv": "X25519",
      "x": "s6-DsrCQcLyd0VOpBCfeOCY4ejcoP6rPqFKGmMJ2Nlo"
    },
    "privateKeyJwk": {
      "kty": "OKP",
      "crv": "X25519",
      "x": "s6-DsrCQcLyd0VOpBCfeOCY4ejcoP6rPqFKGmMJ2Nlo",
      "d": "mMnU-7mxUZqtJ3ndtN4zYi3oqRpz1IphuLBUTPoUk0E"
    }
  }
]
```

**Comment by decentralgabe** [comment]
@OR13 yes, the `d` found in the second key returns this 

Here is the test I'm running on Go 1.17.6
```
d := "fWGQTFQaEc2Kh-LyJhXlUQtcaktFymPtVfcC5SacRew"
decodedD, err := base64.URLEncoding.DecodeString(d)
assert.NoError(t, err)
assert.True(t, len(decodedD) == ed25519.PrivateKeySize)
privateKey := ed25519.NewKeyFromSeed(decodedD)
```

the first assertion shows `illegal base64 data at input byte 40`

the second shows `Should be true`

**Comment by decentralgabe** [comment]
> hmm possible the test vectors were generated from a faulty implementation (probably my fault)

so, I don't think the test vectors are your fault. I think the test vectors are fine. I think the private key is `d+x` which is documented nowhere (that I can see).

**Comment by decentralgabe** [comment]
In searching for a solution I came across https://github.com/w3c/vc-wg-charter/issues/82 which you opened @OR13 - I think it's really pertinent here! If we expect this suite to be used widely, verifiable credentials to be used widely, and DIDs to be used widely, the least we can ask for is some clear documentation.

**Comment by OR13** [comment]
To be clear a JWK private key is just `d`.

the public components are `x` or `x` and `y` for `kty` `EC`.

Since this repo is only about JWK / JWS... there should be no `d+x` discussed here... since thats not a thing in JWK.

**Comment by decentralgabe** [comment]
> To be clear a JWK private key is just `d`.
> 
> the public components are `x` for `kty` `OKP` or `x` and `y` for `kty` `EC`.
> 
> Since this repo is only about JWK / JWS... there should be no `d+x` discussed here... since thats not a thing in JWK.

Fair point. According to the JWK spec the `d` property should be a base64 url encoded private key. It is not. If there existed clear documentation on how to construct Public and Private Key JWKs perhaps this would help. It also may be true that the test vectors and examples in the spec are faulty.

**Comment by kdenhartog** [comment]
@decentralgabe I'm pretty sure it's failing because of the no-padding problem.

In [section 1.1 of RFC7517](https://datatracker.ietf.org/doc/html/rfc7517#section-1.1) it says use the base64url encoding defined in JWS. In the JWS spec, [RFC7515 section 2](https://datatracker.ietf.org/doc/html/rfc7515#section-2), it's expected that base64url encoding in the JOSE suite should drop the padding as defined in [appendix C](https://datatracker.ietf.org/doc/html/rfc7515#appendix-C) of RFC 7515.

In other words, JWKs don't use standard base64url encoding and instead use base64url encoding without padding which is slightly different and why it's erroring. If you add the padding back on then I suspect it will properly base64url decode.

**Comment by decentralgabe** [comment]
> @decentralgabe I'm pretty sure it's failing because of the no-padding problem.
> 
> In [section 1.1 of RFC7517](https://datatracker.ietf.org/doc/html/rfc7517#section-1.1) it says use the base64url encoding defined in JWS. In the JWS spec, [RFC7515 section 2](https://datatracker.ietf.org/doc/html/rfc7515#section-2), it's expected that base64url encoding in the JOSE suite should drop the padding as defined in [appendix C](https://datatracker.ietf.org/doc/html/rfc7515#appendix-C) of RFC 7515.
> 
> In other words, JWKs don't use standard base64url encoding and instead use base64url encoding without padding which is slightly different and why it's erroring. If you add the padding back on then I suspect it will properly base64url decode.

Thank you @kdenhartog. I believe you're right. Using the [example from the spec](https://datatracker.ietf.org/doc/html/rfc8037#appendix-A.1) I run into the same issue.

However, I still think finding this information is far more difficult than it should be. I'd still like to see PublicKeyJWK and PrivateKeyJWK be well-defined.

**Comment by kdenhartog** [comment]
+1 to well defining these things since it's not completely obvious what's expected.

For the specific issue you ran into we could add a short note to point out that publicKeyJwk and privateKeyJwk rely on base64url no padding as well.
